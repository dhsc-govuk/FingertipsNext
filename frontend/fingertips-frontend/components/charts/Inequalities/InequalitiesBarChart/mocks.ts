import { InequalitiesBarChartData } from '@/components/charts/Inequalities/helpers/inequalitiesHelpers';

export const getTestData = (): InequalitiesBarChartData => ({
  areaCode: 'A1425',
  areaName: 'South FooBar',
  data: {
    period: 2008,
    inequalities: {
      Male: {
        value: 890.328253,
        count: 131,
        upper: 578.32766,
        lower: 441.69151,
        isAggregate: false,
      },
      Persons: {
        value: 135.149304,
        count: 222,
        upper: 578.32766,
        lower: 441.69151,
        isAggregate: true,
      },
      Female: {
        value: 890.328253,
        count: 131,
        upper: 578.32766,
        lower: 441.69151,
        isAggregate: false,
      },
    },
  },
});
