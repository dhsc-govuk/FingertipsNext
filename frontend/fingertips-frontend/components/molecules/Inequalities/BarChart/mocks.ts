import { InequalitiesBarChartData } from '@/components/organisms/Inequalities/inequalitiesHelpers';

export const getTestData = (): InequalitiesBarChartData => ({
  areaName: 'South FooBar',
  data: {
    period: 2008,
    inequalities: {
      Male: {
        value: 890.328253,
        count: 131,
        upper: 578.32766,
        lower: 441.69151,
      },
      Persons: {
        value: 135.149304,
        count: 222,
        upper: 578.32766,
        lower: 441.69151,
      },
      Female: {
        value: 890.328253,
        count: 131,
        upper: 578.32766,
        lower: 441.69151,
      },
    },
  },
});
