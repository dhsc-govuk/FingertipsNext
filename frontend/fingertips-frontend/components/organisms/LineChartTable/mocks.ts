export const MOCK_ENGLAND_DATA = {
  areaCode: 'E92000001',
  areaName: 'England',
  healthData: [
    {
      year: 2004,
      count: 200,
      value: 904.874,
      lowerCi: 0,
      upperCi: 0,
      ageBand: '0-4',
      sex: 'Female',
    },
    {
      year: 2008,
      count: 500,
      value: 965.9843,
      lowerCi: 0,
      upperCi: 0,
      ageBand: '10-14',
      sex: 'Female',
    },
  ],
};

export const MOCK_HEALTH_DATA = [
  {
    areaCode: 'A1425',
    areaName: 'Greater Manchester ICB - 00T',
    healthData: [
      {
        year: 2008,
        count: 222,
        value: 890.305692,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Persons',
      },
      {
        year: 2004,
        count: 267,
        value: 703.420759,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Persons',
      },
    ],
  },
  {
    ...MOCK_ENGLAND_DATA,
  },
  {
    areaCode: 'A1426',
    areaName: 'South FooBar',
    healthData: [
      {
        year: 2004,
        count: 222,
        value: 135.149304,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Persons',
      },
      {
        year: 2008,
        count: 131,
        value: 890.328253,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Persons',
      },
    ],
  },
];
