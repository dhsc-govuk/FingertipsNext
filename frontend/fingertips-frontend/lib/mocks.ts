import {
  AgeData,
  DeprivationData,
  HealthDataPoint,
  HealthDataPointTrendEnum,
  SexData,
} from '@/generated-sources/ft-api-client';
import { mockDatePeriod } from '@/mock/data/mockDatePeriod';

export const noDeprivation: DeprivationData = {
  sequence: 1,
  type: 'All',
  value: 'All',
  isAggregate: true,
};

export const personsSex: SexData = {
  value: 'Persons',
  isAggregate: true,
};

export const maleSex: SexData = {
  value: 'Male',
  isAggregate: false,
};

export const femaleSex: SexData = {
  value: 'Female',
  isAggregate: false,
};

export const allAgesAge: AgeData = {
  value: 'All ages',
  isAggregate: true,
};

export const disaggregatedAge = (age: string) => {
  return {
    value: age,
    isAggregate: false,
  };
};

export const healthDataPoint: HealthDataPoint = {
  count: 389,
  lowerCi: 441.69151,
  upperCi: 578.32766,
  value: 278.29134,
  datePeriod: mockDatePeriod(2006),
  sex: personsSex,
  ageBand: allAgesAge,
  trend: HealthDataPointTrendEnum.NotYetCalculated,
  deprivation: noDeprivation,
};
