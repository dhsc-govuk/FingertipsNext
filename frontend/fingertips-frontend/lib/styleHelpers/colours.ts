export enum GovukColours {
  Red = '#D4351C',
  Yellow = '#FFDD00',
  Green = '#00703C',
  Blue = '#1D70B8',
  DarkBlue = '#003078',
  LightBlue = '#5694CA',
  Purple = '#4C2C92',
  Black = '#0B0C0C',
  DarkGrey = '#505A5F',
  MidGrey = '#B1B4B6',
  LightGrey = '#F3F2F1',
  White = '#FFFFFF',
  LightPurple = '#A285D1',
  BrightPurple = '#912B88',
  Pink = '#D53880',
  LightPink = '#F499BE',
  Orange = '#F46A25',
  Brown = '#B58840',
  LightGreen = '#85994B',
  Turquoise = '#28A197',
  DarkPink = '#801650',
  OtherLightPurple = '#6F72AF',
  CharcoalGray = '#3D3D3D',
  Female = '#5352BE',
  Male = '#57AEF8',
  DarkSlateGray = '#D7D7D7',
}

// govuk-react table border colour
// it is the old gds grey 2
// and it appears govuk-react hasn't been updated with the latest gds colours
export const BorderColour = '#BfC1C3';

export enum TagColours {
  GreyText = '#303436',
  GreyBackground = '#E1E2E3',
  // benchmarks
  DarkRed = '#831504',
  LightBlue = '#AFCFE5', //chart-lower99
  // trends
  YellowBackground = '#FFF5BB',
  YellowText = '#514311',
  GreenBackground = '#C4DED3',
  GreenText = '#00502D',
  RedBackground = '#F6C6BF',
  RedText = '#270D0B',
  LightBlueBackground = '#E3EFF6',
}

export enum QuintileColours {
  Lowest = '#E4DDFF',
  Low = '#CBBEF4',
  Middle = '#AA90EC',
  High = '#8B60E2',
  Highest = '#6B33C3',
  Worst = '#D494C1',
  Worse = '#BC6AAA',
  MiddleWithValue = '#A44596',
  Better = '#812972',
  Best = '#561950',
}

export function getTextColour(backgroundColour: string) {
  switch (backgroundColour) {
    case GovukColours.Green:
    case GovukColours.Red:
    case GovukColours.LightBlue:
    case GovukColours.DarkBlue:
    case GovukColours.Blue:
    case GovukColours.Black:

    case TagColours.DarkRed:

    case QuintileColours.High:
    case QuintileColours.Highest:
    case QuintileColours.MiddleWithValue:
    case QuintileColours.Better:
    case QuintileColours.Best:
      return GovukColours.White;

    default:
      return GovukColours.Black;
  }
}
