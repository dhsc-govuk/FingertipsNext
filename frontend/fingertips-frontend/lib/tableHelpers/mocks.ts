export const MOCK_ENGLAND_DATA = {
  areaCode: 'E92000001',
  areaName: 'England',
  healthData: [
    {
      year: 2004,
      count: 200,
      value: 904.874,
      lowerCi: 0,
      upperCi: 0,
      ageBand: '0-4',
      sex: 'All',
    },
    {
      year: 2008,
      count: 500,
      value: 965.9843,
      lowerCi: 0,
      upperCi: 0,
      ageBand: '10-14',
      sex: 'All',
    },
  ],
};

export const MOCK_HEALTH_DATA = [
  {
    areaCode: 'A1425',
    areaName: 'Greater Manchester ICB - 00T',
    healthData: [
      {
        year: 2008,
        count: 222,
        value: 890.305692,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'All',
      },
      {
        year: 2004,
        count: 267,
        value: 703.420759,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'All',
      },
      {
        year: 2004,
        count: 267,
        value: 703.420759,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Male',
      },
      {
        year: 2004,
        count: 267,
        value: 703.420759,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Female',
      },
    ],
  },
  {
    areaCode: 'A1426',
    areaName: 'South FooBar',
    healthData: [
      {
        year: 2008,
        count: 222,
        value: 135.149304,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'All',
      },
      {
        year: 2008,
        count: 131,
        value: 890.328253,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Male',
      },
      {
        year: 2008,
        count: 131,
        value: 890.328253,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        ageBand: 'All',
        sex: 'Female',
      },
    ],
  },
];

export const MOCK_PARENT_DATA = {
  areaCode: 'E40000014',
  areaName: 'NHS North West Region',
  healthData: [
    {
      year: 2008,
      count: 777,
      value: 1000,
      lowerCi: 500,
      upperCi: 1500,
      ageBand: 'All',
      sex: 'All',
    },
    {
      year: 2004,
      count: 777,
      value: 1000,
      lowerCi: 500,
      upperCi: 1500,
      ageBand: 'All',
      sex: 'All',
    },
  ],
};
