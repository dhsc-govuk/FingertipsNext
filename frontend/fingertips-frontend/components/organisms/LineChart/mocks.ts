import {
  HealthDataForArea,
  HealthDataPointTrendEnum,
} from '@/generated-sources/ft-api-client';
import { allAgesAge, noDeprivation, personsSex } from '@/lib/mocks';

export const mockIndicatorData: HealthDataForArea[] = [
  {
    areaCode: 'A1425',
    areaName: 'North FooBar',
    healthData: [
      {
        count: 389,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        value: 278.29134,
        year: 2006,
        periodLabel: '2006',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
      {
        count: 267,
        lowerCi: 441.69151,
        upperCi: 578.32766,
        value: 703.420759,
        year: 2004,
        periodLabel: '2004',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
    ],
  },
  {
    areaCode: 'A1426',
    areaName: 'South FooBar',
    healthData: [
      {
        count: 654,
        lowerCi: 750.69151,
        upperCi: 800.32766,
        value: 786.27434,
        year: 2010,
        periodLabel: '2010',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
      {
        count: 567,
        lowerCi: 440.69151,
        upperCi: 420.32766,
        value: 435.420759,
        year: 2007,
        periodLabel: '2007',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
    ],
  },
  {
    areaCode: 'A1427',
    areaName: 'East FooBar',
    healthData: [
      {
        count: 953,
        lowerCi: 460.69151,
        upperCi: 500.32766,
        value: 478.27434,
        year: 2020,
        periodLabel: '2020',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
      {
        count: 567,
        lowerCi: 220.69151,
        upperCi: 250.32766,
        value: 234.420759,
        year: 2012,
        periodLabel: '2012',
        sex: personsSex,
        ageBand: allAgesAge,
        trend: HealthDataPointTrendEnum.NotYetCalculated,
        deprivation: noDeprivation,
        isAggregate: true,
      },
    ],
  },
];

export const mockBenchmarkData: HealthDataForArea = {
  areaCode: 'E92000001',
  areaName: 'England',
  healthData: [
    {
      count: 389,
      lowerCi: 441.69151,
      upperCi: 578.32766,
      value: 278.29134,
      year: 2006,
      periodLabel: '2006',
      sex: personsSex,
      ageBand: allAgesAge,
      trend: HealthDataPointTrendEnum.NotYetCalculated,
      deprivation: noDeprivation,
      isAggregate: true,
    },
    {
      count: 267,
      lowerCi: 441.69151,
      upperCi: 578.32766,
      value: 703.420759,
      year: 2004,
      periodLabel: '2004',
      sex: personsSex,
      ageBand: allAgesAge,
      trend: HealthDataPointTrendEnum.NotYetCalculated,
      deprivation: noDeprivation,
      isAggregate: true,
    },
  ],
};

export const mockParentData: HealthDataForArea = {
  areaCode: 'P001',
  areaName: 'Parent',
  healthData: [
    {
      count: 100,
      lowerCi: 200,
      upperCi: 400,
      value: 300,
      year: 2006,
      periodLabel: '2006',
      sex: personsSex,
      ageBand: allAgesAge,
      trend: HealthDataPointTrendEnum.NotYetCalculated,
      deprivation: noDeprivation,
      isAggregate: true,
    },
    {
      count: 101,
      lowerCi: 201,
      upperCi: 401,
      value: 301,
      year: 2004,
      periodLabel: '2004',
      sex: personsSex,
      ageBand: allAgesAge,
      trend: HealthDataPointTrendEnum.NotYetCalculated,
      deprivation: noDeprivation,
      isAggregate: true,
    },
  ],
};
