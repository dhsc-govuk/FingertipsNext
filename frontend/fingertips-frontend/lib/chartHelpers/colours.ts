import { GovukColours } from '../styleHelpers/colours';

export type ChartColours = GovukColours | UniqueChartColours;

export enum UniqueChartColours {
  ChartDarkBlue = '#12436D',
  OtherLightBlue = '#57AEF8',
}

// The ordering of this array determines the order in which the colours are
// used by Highcharts, so it should be preserved.
export const chartColours: ChartColours[] = [
  GovukColours.Yellow,
  GovukColours.Red,
  GovukColours.Blue,
  GovukColours.LightPink,
  GovukColours.Brown,
  UniqueChartColours.ChartDarkBlue,
  GovukColours.LightBlue,
  GovukColours.OtherLightPurple,
  GovukColours.Orange,
  GovukColours.LightPurple,
  GovukColours.DarkPink,
  GovukColours.Green,
  GovukColours.Pink,
  GovukColours.Purple,
];
