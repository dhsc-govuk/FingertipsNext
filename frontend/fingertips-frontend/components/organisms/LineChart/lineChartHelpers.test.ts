import { SeriesLineOptions, SymbolKeyValue } from 'highcharts';
import {
  generateSeriesData,
  generateStandardLineChartOptions,
  lineChartDefaultOptions,
} from './lineChartHelpers';
import { GovukColours } from '@/lib/styleHelpers/colours';
import { mockIndicatorData, mockBenchmarkData, mockParentData } from './mocks';

const symbols: SymbolKeyValue[] = ['arc', 'circle', 'diamond'];

const chartColours: GovukColours[] = [
  GovukColours.Orange,
  GovukColours.LightPurple,
  GovukColours.DarkPink,
];

describe('generateSeriesData', () => {
  it('should generate series data without benchmark data', () => {
    const expectedSeriesData = [
      {
        color: '#F46A25',
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2006, 441.69151, 578.32766],
          [2004, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#A285D1',
        data: [
          [2010, 786.27434],
          [2007, 435.420759],
        ],
        name: 'South FooBar',
        type: 'line',
        marker: {
          symbol: 'circle',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2010, 750.69151, 800.32766],
          [2007, 440.69151, 420.32766],
        ],
        name: 'South FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#801650',
        data: [
          [2020, 478.27434],
          [2012, 234.420759],
        ],
        name: 'East FooBar',
        type: 'line',
        marker: {
          symbol: 'diamond',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2020, 460.69151, 500.32766],
          [2012, 220.69151, 250.32766],
        ],
        name: 'East FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    const generatedSeriesData = generateSeriesData(
      mockIndicatorData,
      symbols,
      chartColours,
      undefined,
      undefined,
      false
    );

    expect(generatedSeriesData).toEqual(expectedSeriesData);
  });

  it('should generate series data with benchmark data', () => {
    const expectedSeriesData = [
      {
        color: GovukColours.DarkGrey,
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        marker: {
          symbol: 'circle',
        },
        name: 'Benchmark: England',
        type: 'line',
      },
      {
        color: '#F46A25',
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2006, 441.69151, 578.32766],
          [2004, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#A285D1',
        data: [
          [2010, 786.27434],
          [2007, 435.420759],
        ],
        name: 'South FooBar',
        type: 'line',
        marker: {
          symbol: 'circle',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2010, 750.69151, 800.32766],
          [2007, 440.69151, 420.32766],
        ],
        name: 'South FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#801650',
        data: [
          [2020, 478.27434],
          [2012, 234.420759],
        ],
        name: 'East FooBar',
        type: 'line',
        marker: {
          symbol: 'diamond',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2020, 460.69151, 500.32766],
          [2012, 220.69151, 250.32766],
        ],
        name: 'East FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    const generatedSeriesData = generateSeriesData(
      mockIndicatorData,
      symbols,
      chartColours,
      mockBenchmarkData,
      undefined,
      false
    );

    expect(generatedSeriesData).toEqual(expectedSeriesData);
  });

  it('should generate series data with parent data', () => {
    const expectedSeriesData = [
      {
        color: GovukColours.DarkGrey,
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        marker: {
          symbol: 'circle',
        },
        name: 'Benchmark: England',
        type: 'line',
      },
      {
        color: GovukColours.Turquoise,
        data: [
          [2006, 300],
          [2004, 301],
        ],
        marker: {
          symbol: 'diamond',
        },
        name: 'Group: Parent',
        type: 'line',
      },
      {
        color: '#F46A25',
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2006, 441.69151, 578.32766],
          [2004, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#A285D1',
        data: [
          [2010, 786.27434],
          [2007, 435.420759],
        ],
        name: 'South FooBar',
        type: 'line',
        marker: {
          symbol: 'circle',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2010, 750.69151, 800.32766],
          [2007, 440.69151, 420.32766],
        ],
        name: 'South FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
      {
        color: '#801650',
        data: [
          [2020, 478.27434],
          [2012, 234.420759],
        ],
        name: 'East FooBar',
        type: 'line',
        marker: {
          symbol: 'diamond',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2020, 460.69151, 500.32766],
          [2012, 220.69151, 250.32766],
        ],
        name: 'East FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    const generatedSeriesData = generateSeriesData(
      mockIndicatorData,
      symbols,
      chartColours,
      mockBenchmarkData,
      mockParentData,
      false
    );

    expect(generatedSeriesData).toEqual(expectedSeriesData);
  });

  it('should repeat symbols when there are more series than symbols', () => {
    const symbols: SymbolKeyValue[] = ['arc', 'circle'];

    const generatedSeriesData = generateSeriesData(
      mockIndicatorData,
      symbols,
      chartColours,
      undefined,
      undefined,
      false
    ) as SeriesLineOptions[];

    expect(generatedSeriesData[0].marker?.symbol).toBe('arc');
    // The error bar series has no symbol hence undefined
    expect(generatedSeriesData[1].marker?.symbol).toBe(undefined);
    expect(generatedSeriesData[2].marker?.symbol).toBe('circle');
    expect(generatedSeriesData[3].marker?.symbol).toBe(undefined);
    expect(generatedSeriesData[4].marker?.symbol).toBe('arc');
  });

  it('should show confidence intervals bars', () => {
    const generatedSeriesData = generateSeriesData(
      [mockIndicatorData[0]],
      symbols,
      chartColours,
      undefined,
      undefined,
      true
    );

    const expectedSeriesData = [
      {
        color: '#F46A25',
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2006, 441.69151, 578.32766],
          [2004, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: true,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    expect(generatedSeriesData).toEqual(expectedSeriesData);
  });

  it('should not show confidence intervals bars', () => {
    const generatedSeriesData = generateSeriesData(
      [mockIndicatorData[0]],
      symbols,
      chartColours,
      undefined,
      undefined,
      false
    );

    const expectedSeriesData = [
      {
        color: '#F46A25',
        data: [
          [2006, 278.29134],
          [2004, 703.420759],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2006, 441.69151, 578.32766],
          [2004, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    expect(generatedSeriesData).toEqual(expectedSeriesData);
  });

  it('should repeat colours when there are more series than colours', () => {
    const chartColours: GovukColours[] = [
      GovukColours.Orange,
      GovukColours.LightPurple,
    ];
    const errorBarColour = '#B1B4B6';

    const generatedSeriesData = generateSeriesData(
      mockIndicatorData,
      symbols,
      chartColours,
      undefined,
      undefined,
      false
    ) as SeriesLineOptions[];

    expect(generatedSeriesData[0].color).toBe(chartColours[0]);
    expect(generatedSeriesData[1].color).toBe(errorBarColour);
    expect(generatedSeriesData[2].color).toBe(chartColours[1]);
    expect(generatedSeriesData[3].color).toBe(errorBarColour);
    expect(generatedSeriesData[4].color).toBe(chartColours[0]);
  });
});

describe('generateStandardLineChartOptions', () => {
  it('should generate standard line chart options', () => {
    const expectedSeriesData = [
      {
        color: '#F46A25',
        data: [
          [2004, 703.420759],
          [2006, 278.29134],
        ],
        name: 'North FooBar',
        type: 'line',
        marker: {
          symbol: 'arc',
        },
      },
      {
        color: '#B1B4B6',
        data: [
          [2004, 441.69151, 578.32766],
          [2006, 441.69151, 578.32766],
        ],
        name: 'North FooBar',
        type: 'errorbar',
        visible: false,
        lineWidth: 2,
        whiskerLength: '20%',
      },
    ];

    const expected = {
      ...lineChartDefaultOptions,
      yAxis: {
        ...lineChartDefaultOptions.yAxis,
        title: { text: 'yAxis', margin: 20 },
      },
      xAxis: {
        ...lineChartDefaultOptions.xAxis,
        title: { text: 'xAxis', margin: 20 },
      },
      legend: {
        ...lineChartDefaultOptions.legend,
        title: { text: 'Areas' },
      },
      accessibility: {
        ...lineChartDefaultOptions.accessibility,
        description: 'accessibility',
      },
      series: expectedSeriesData,
      tooltip: {
        format: lineChartDefaultOptions.tooltip?.format + '%',
      },
    };

    expect(
      generateStandardLineChartOptions([mockIndicatorData[0]], false, {
        benchmarkData: undefined,
        groupIndicatorData: undefined,
        yAxisTitle: 'yAxis',
        xAxisTitle: 'xAxis',
        measurementUnit: '%',
        accessibilityLabel: 'accessibility',
        colours: chartColours,
        symbols,
      })
    ).toEqual(expected);
  });

  it('should generate standard line chart options with benchmark data', () => {
    expect(
      generateStandardLineChartOptions([mockIndicatorData[0]], false, {
        benchmarkData: mockBenchmarkData,
        groupIndicatorData: mockParentData,
        yAxisTitle: 'yAxis',
        xAxisTitle: 'xAxis',
        measurementUnit: '%',
        accessibilityLabel: 'accessibility',
        colours: chartColours,
        symbols,
      })
    ).toMatchSnapshot();
  });
});
