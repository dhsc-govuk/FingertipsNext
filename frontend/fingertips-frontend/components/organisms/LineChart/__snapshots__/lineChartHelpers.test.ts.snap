// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateStandardLineChartOptions should generate standard line chart options 1`] = `
{
  "accessibility": {
    "description": "accessibility",
    "enabled": false,
  },
  "chart": {
    "animation": false,
    "spacingBottom": 50,
    "spacingTop": 20,
    "type": "line",
  },
  "credits": {
    "enabled": false,
  },
  "legend": {
    "align": "left",
    "itemStyle": {
      "fontSize": "16px",
    },
    "margin": 20,
    "verticalAlign": "top",
  },
  "plotOptions": {
    "series": {
      "animation": false,
    },
  },
  "series": [
    {
      "color": "#F46A25",
      "data": [
        [
          2004,
          703.420759,
        ],
        [
          2006,
          278.29134,
        ],
      ],
      "marker": {
        "symbol": "arc",
      },
      "name": "North FooBar",
      "type": "line",
    },
    {
      "color": "#B1B4B6",
      "data": [
        [
          2004,
          441.69151,
          578.32766,
        ],
        [
          2006,
          441.69151,
          578.32766,
        ],
      ],
      "lineWidth": 2,
      "name": "North FooBar",
      "type": "errorbar",
      "visible": false,
      "whiskerLength": "20%",
    },
  ],
  "title": {
    "style": {
      "display": "none",
    },
  },
  "tooltip": {
    "formatter": [Function],
  },
  "xAxis": {
    "allowDecimals": false,
    "labels": {
      "style": {
        "fontSize": 16,
      },
    },
    "tickLength": 0,
    "title": {
      "margin": 20,
      "style": {
        "fontSize": 19,
      },
      "text": "xAxis",
    },
  },
  "yAxis": {
    "labels": {
      "style": {
        "fontSize": 16,
      },
    },
    "minorTickInterval": "auto",
    "minorTicksPerMajor": 2,
    "title": {
      "margin": 20,
      "style": {
        "fontSize": 19,
      },
      "text": "yAxis",
    },
  },
}
`;

exports[`generateStandardLineChartOptions should generate standard line chart options with benchmark data 1`] = `
{
  "accessibility": {
    "description": "accessibility",
    "enabled": false,
  },
  "chart": {
    "animation": false,
    "spacingBottom": 50,
    "spacingTop": 20,
    "type": "line",
  },
  "credits": {
    "enabled": false,
  },
  "legend": {
    "align": "left",
    "itemStyle": {
      "fontSize": "16px",
    },
    "margin": 20,
    "verticalAlign": "top",
  },
  "plotOptions": {
    "series": {
      "animation": false,
    },
  },
  "series": [
    {
      "color": "#505A5F",
      "data": [
        [
          2004,
          703.420759,
        ],
        [
          2006,
          278.29134,
        ],
      ],
      "marker": {
        "symbol": "circle",
      },
      "name": "Benchmark: England",
      "type": "line",
    },
    {
      "color": "#28A197",
      "data": [
        [
          2004,
          301,
        ],
        [
          2006,
          300,
        ],
      ],
      "marker": {
        "symbol": "diamond",
      },
      "name": "Group: Parent",
      "type": "line",
    },
    {
      "color": "#F46A25",
      "data": [
        [
          2004,
          703.420759,
        ],
        [
          2006,
          278.29134,
        ],
      ],
      "marker": {
        "symbol": "arc",
      },
      "name": "North FooBar",
      "type": "line",
    },
    {
      "color": "#B1B4B6",
      "data": [
        [
          2004,
          441.69151,
          578.32766,
        ],
        [
          2006,
          441.69151,
          578.32766,
        ],
      ],
      "lineWidth": 2,
      "name": "North FooBar",
      "type": "errorbar",
      "visible": false,
      "whiskerLength": "20%",
    },
  ],
  "title": {
    "style": {
      "display": "none",
    },
  },
  "tooltip": {
    "formatter": [Function],
  },
  "xAxis": {
    "allowDecimals": false,
    "labels": {
      "style": {
        "fontSize": 16,
      },
    },
    "tickLength": 0,
    "title": {
      "margin": 20,
      "style": {
        "fontSize": 19,
      },
      "text": "xAxis",
    },
  },
  "yAxis": {
    "labels": {
      "style": {
        "fontSize": 16,
      },
    },
    "minorTickInterval": "auto",
    "minorTicksPerMajor": 2,
    "title": {
      "margin": 20,
      "style": {
        "fontSize": 19,
      },
      "text": "yAxis",
    },
  },
}
`;
